### ISSUE TITLE
Multi-Level BOM Display

### ISSUE THREAD

**Issue by SchrodingersGat** [inventree/InvenTree#745]
In the "Bill of Materials" display, we already show which "BOM items" are assemblies themselves. Here, we should allow the user to "expand" the row and show the sub items under that BOM.

This should cascade down to the lowest level available.

**Comment by SchrodingersGat** [comment]
Ref: https://bootstrap-table.com/docs/extensions/treegrid/

**Comment by eeintech** [comment]
@SchrodingersGat I'm looking for a better implementation of BOMs and export. I see there is a bunch of open issue on this topic:
https://github.com/inventree/InvenTree/issues?q=is%3Aissue+is%3Aopen+BOM

Do you need help with this? I'll be happy to give it a shot.

**Comment by SchrodingersGat** [comment]
> Do you need help with this? I'll be happy to give it a shot.

Absolutely - my bandwidth is pretty reduced at the moment with child and work. What exactly do you have in mind? :)

**Comment by eeintech** [comment]
It seems to me that a good implementation for BOM flow relies upon:
* An import process that can both manual and automated (to connect to CAD software -> REST API or Python Module should do) and indicate/report the current level of stocks (#760) 
* A showing view (using treegrid) that allows adding, allocating and editing BOM items after import (#745) with a way of locking a BOM to prevent other users to alter it
* An export tool that takes into account all suppliers/manufacturers parts (eg. alternates #841 and their selection at the time of the build) and let user select which fields to export (#793, #126)

I added links to current issues reported by users (and you obviously), I think there is a lot more to it once digging deeper but this is my overall impression. Happy to hear your side. Ultimately, I would love to see this section of the docs filling up with good stuff: https://inventree.github.io/docs/build/bom

On a side note, most part management systems (PLM, ERP) will have a dedicated "section" for BOMs and release process. I'm fine with the current implementation, I just wanted to point out BOM flow holds a huge role as it drives successful builds and part selection, the importance of the BOM can't be minimized. One could consider it essential for having a tab/section dedicated to BOM in InvenTree.

**Comment by SchrodingersGat** [comment]
The BOM workflow is probably the area that:

a) Needs the most work
b) I am the least happy with

TL;DR - I would *love* some help with all the issues you point out above. If you need some pointers on any particular issues LMK.

## Treegrid

Currently all tables are rendered using bootstrap-table - there's a treegrid plugin which is documented here  - https://bootstrap-table.com/docs/extensions/treegrid/

**Comment by eeintech** [comment]
Alright then, I'm gonna focus on this in the scope of my work and following the lines I pointed out but will try as much as possible to keep it intuitive like the rest of InvenTree. I'm no master in Django (aspiring to, I wrote several apps and can read and understand InvenTree's code) and front-end (maybe one day) but I have a deep understanding of proper usage of BoMs with my past and current Electrical Engineer experiences. If you have any pointers for code formatting, flow, etc. don't hesitate, else I'll try to stay true to the Python/Django/InvenTree spirit based on the current implementation.
I won't be able to work full-time on this but will try my best to get this going soon!

**Comment by SchrodingersGat** [comment]
I'm very excited to see what you come up with. In particular a different perspective on how to manage BOM's is going to be fantastic. :)

**Comment by SchrodingersGat** [comment]
@eeintech have you managed to have any time to look into this? :)

**Comment by eeintech** [comment]
@SchrodingersGat I've looked up the current implementation but haven't had the chance to put any time into implementing it yet. I should have soon, it's on my TODO list 😃
